### PR TITLE
Demo & Dataset work

### PR DESCRIPTION
**Pull request by stackus**
@AnkyrinRepeat, @OBCENEIKON 

I had to recreate this pull request because the previous one was merged into master and reverted it. Please fill in comments below with what is in this PR.

### CODE REVIEW

File: app/models/jellyfish_demo/service/storage.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          image_id: product.settings[:image_id],
+          flavor_id: product.settings[:flavor_id],
+          key_name: product.settings[:key_name],
+          subnet_id: product.settings[:subnet_id],
+          instance_id: random_id,
+          public_ip_address: "192.178.0.#{random_ip}"
+        }
+
+        # SAVE PRODUCT DETAILS
+        save_outputs(details, [['image_id', :image_id], ['flavor_id', :flavor_id], ['key_name', :key_name],
```
Comment (stackus) [inline app/models/jellyfish_demo/service/storage.rb]: All of the strings match the symbols. You can get a string from a symbol with `to_s`, and get a symbol from a string with `to_sym`.

You're using `defined?` for a value that was declared one statement above. This was copy and pasted from elsewhere?

`if defined? details` isn't doing what you think.

``` ruby
defined? 'foobar'  # "expression"
defined? nil       # "nil"
defined? true      # "true"
defined? false     # "false"
defined? TrueClass # "constant"
```

File: app/models/jellyfish_demo/product_type/software.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (stackus) [inline app/models/jellyfish_demo/product_type/software.rb]: Blank lines at the start of many files including this one.

File: db/data/sample/groups.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+---
```
Comment (stackus) [inline db/data/sample/groups.yml]: This sample data, not just this file, should be removed. The method in which demo data is installed should be automatic.

The data and task need to be refactored and cannot be kept as is.

Comment (AnkyrinRepeat) [inline db/data/sample/groups.yml]: Should that be done with a seeds.rb file?

File: app/models/jellyfish_demo/service/storage.rb (hunk 1 above)
Comment (AnkyrinRepeat) [inline app/models/jellyfish_demo/service/storage.rb]: It was copy/pasted from jellyfish_aws.

Should I replace it with `save_outputs(details, ['image_id', 'flavor_id', 'key_name', 'subnet_id', 'instance_id', 'public_ip_address'], ValueTypes::TYPES[:string])`?

File: db/data/sample/groups.yml (hunk 3 above)
Comment (stackus) [inline db/data/sample/groups.yml]: "automatic" was the wrong word.

The idea is that three sets of data would be created initially. The 'Federal', 'Health', and 'generic' sets. These datasets should be able to be installed without a rake task.

Ideally, the UI used to installed them can inform the admin user if a dataset has been installed and which one, and also provide a way to uninstall the dataset (pie in the sky stretch).

File: app/models/jellyfish_demo/service/storage.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        random_id = 1 + rand(99_999)
+        random_ip = rand(999)
+
+        details = {
+          image_id: product.settings[:image_id],
+          flavor_id: product.settings[:flavor_id],
+          key_name: product.settings[:key_name],
+          subnet_id: product.settings[:subnet_id],
+          instance_id: random_id,
+          public_ip_address: "192.178.0.#{random_ip}"
```
Comment (stackus) [inline app/models/jellyfish_demo/service/storage.rb:19]: `192.178.0.999` is not a valid IP address.

File: app/models/jellyfish_demo/service/storage.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
+module JellyfishDemo
+  module Service
+    class Storage < ::Service
+      def actions
+        actions = super.merge :terminate
+        actions
+      end
+
+      def provision
+        random_id = 1 + rand(99_999)
```
Comment (stackus) [inline app/models/jellyfish_demo/service/storage.rb:10]: `rand 2..99_999` and the next line `rand 1..254`

Ruby has `Ranges`, these can be defined as two values with either `..` or `...` between them. http://ruby-doc.org/core-2.3.0/Range.html

File: app/models/jellyfish_demo/service/storage.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        update_status :terminated, 'terminated'
+      end
+
+      private
+
+      def update_status(status, status_msg)
+        update_attributes status: status, status_msg: status_msg
+      end
+
+      def save_outputs(source, outputs_to_save, output_value_type)
```
Comment (stackus) [inline app/models/jellyfish_demo/service/storage.rb]: You will want to reevaluate if you need this method for the demo provisioning. It also uses `defined?`

**Pull request closed without merging**